src/jit/namedintrinsiclist.h

```diff
@@ -7,7 +7,7 @@
 
 // Named jit intrinsics
 
-enum NamedIntrinsic : unsigned int
```

Comment:
Why was this changed?

Reply:
`unsigned int` size is exceedingly big to code for intrinsics - we need at most 16 bits - it is more than 65 000 names which can be coded. It is more than enough to encode all current and several future generations of Intel and Arm intrinsics. 

Part of the C++ intrinsics names corresponds to method signature - exactly to return and argument types - it is lost during translation to NamedIntrinsic enum values, since they do not code this info.. Furthermore, full translation of C++ intrinsics into emitted code in several situation requires arguments type information  i.e. in case of conversion functions:  [`NI_SSE2_ConvertToVector128Int32`](https://github.com/dotnetrt/coreclr/blob/6c767adc818b672119ed2060d74e4c927de410af/src/jit/hwintrinsiccodegenxarch.cpp#L482) and other similar. I have assumed that we should save space in GenTree nodes and placed type info into 2 bytes besides `NamedIntrinsic enum` which size was shrinked to 2 bytes. This way it allows to read whole intrinsic type info including argument types into `unsigned int` in single operation and use it later with appropriate masking.

I would like to propose using this 2 bytes even more effectively by encoding type info of up to 3 arguments, including immediate values. This would require squeezing it into IMO two 6 bits values for arg type and one 4 bit value for immediate type - or even go for 4 argument types with 4 bit representation each. It allows to encode all numeric types, string and some of IEE 754 binary types.

Reply:
I don't think that is actually needed. The only case where we lose all type information is for the StaticCast Intrinsics, which are <T, U>. 

For all other cases we either have no type information in the Named Intrinsic, but both source operand and return type are consistent, or we have something like the Convert Intrinsics, which have the return type in the intrinsic name.

In both cases, we should just be able to use the existing base Type field to carry the information we lose, and assert/assume the other. (I'm not at a computer right now  so I'll confirm when I get back, but I'm pretty sure this is true for all the Intrinsics we expose right now)

Reply:
@tannergooding 
This info could be useful here as well: https://github.com/tannergooding/coreclr/blob/1e9ea6bb831cc93724470aae8c5d48f6f5e4152b/src/jit/lsraxarch.cpp#L2565

Reply:
Yes, but I haven't seen any intrinsics for which the current tree can't carry the required information.

We basically have the following intrinsic types:

Those like `Add` or `Sqrt` which are `Vector<T> Name(Vector<T>, [Vector<T>, ...])`
* baseType can be T

Those like `Load` or `Set` which are `Vector<T> Name(T, [T, ...])`
* baseType can be T

Those like `CompareGreaterThanOrderedScalar`, which are `bool Name(Vector<T>, Vector<T>)`
* baseType can be T
* return type is known by intrinsic name
* intrinsics required special codegen

Those like `ConvertToInt32`, which are `U NameU(Vector<T>)` or like `Vector<U> NameVectorU(Vector<T>)`
* baseType can be T
* return type (U) is known by intrinsic name

Those like `ConvertToVector128SingleScalar`, which are `Vector<T> NameVectorT(Vector<T>, U)`
* baseType can be U
* returnType (T) is known by intrinsic name

Those like `StaticCast`, which are `Vector<U> Name(Vector<U>)`
* baseType can only be T or U, but
* validation is done in the compiler layer
* requires special or no codegen

Those like `Shuffle`, which are `Vector<T> Name(Vector<T>, Vector<T>, imm)
* baseType can be U
* immediate is just carried by the GenTree opNode

So, in all cases, we can carry the required information between `intrinsicID` and `baseType` (where, what exactly base type represents might vary from intrinsic to intrinsic, but is consistent for a given signature type).

We also don't need to carry anything like the implicit `ival` for the `Compare` operations since those can be pulled from a `HWIntrinsicsInfo` table (something like https://github.com/dotnet/coreclr/pull/15642, although this is going to wait until people are back from the holidays so we can further discuss what fields the table needs -- the current table works fairly well, but there are improvements that could be had to make containment analysis and the like easier, IMO)